### PR TITLE
feat(core): add isPromise generic

### PR DESCRIPTION
**Pull request by danranVm**
add generic to help type inference
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/common/upgrade/src/utils.ts [hunk 1]
```diff
@@ -31,7 +31,7 @@ export function isAnchor(el: (Node & ParentNode) | Element | null): el is HTMLAn
   return (<HTMLAnchorElement>el).href !== undefined;
 }
 
-export function isPromise(obj: any): obj is Promise<any> {
+export function isPromise<T = any>(obj: any): obj is Promise<T> {
   // allow any Promise/A+ compliant thenable.
   // It's up to the caller to ensure that obj.then conforms to the spec
   return !!obj && typeof obj.then === 'function';
```
Comment (danranVm) [inline packages/common/upgrade/src/utils.ts]: Is this better, but I worry about breaking change.
```suggestion
  return !!obj && typeof obj.then === 'function' && typeof obj.catch === 'function';
```

Comment (alfaproject) [reply packages/common/upgrade/src/utils.ts]: You don't need to check for the `catch`

Comment (danranVm) [reply packages/common/upgrade/src/utils.ts]: Okay

File: packages/core/src/application_ref.ts [hunk 2]
```diff
@@ -406,7 +406,7 @@ function _callAndReportToErrorHandler(
   try {
     const result = callback();
     if (isPromise(result)) {
-      return result.catch((e: any) => {
+      return (result as Promise<any>).catch((e: any) => {
```
Comment (petebacondarwin) [inline packages/core/src/application_ref.ts]: Could you not do:

```ts
    if (isPromise<any>(result)) {
	      return result.catch((e: any) => {
```

File: packages/core/src/application_ref.ts [hunk 3]
```diff
@@ -406,7 +406,7 @@ function _callAndReportToErrorHandler(
   try {
     const result = callback();
     if (isPromise(result)) {
-      return (result as Promise<any>).catch((e: any) => {
+      return result.catch((e: any) => {
```
Comment (petebacondarwin) [inline packages/core/src/application_ref.ts]: 👍

File: packages/core/testing/src/testing_internal.ts [hunk 4]
```diff
@@ -138,7 +138,7 @@ function _it(jsmFn: Function, testName: string, testFn: TestFn, testTimeout = 0)
     if (testFn.length === 0) {
       // TypeScript doesn't infer the TestFn type without parameters here, so we
       // need to manually cast it.
-      const retVal = (testFn as () => any)();
+      const retVal = (testFn as() => any)();
```
Comment (petebacondarwin) [inline packages/core/testing/src/testing_internal.ts]: These lint errors must surely be a bug in the linter/formatter?

**Review by petebacondarwin — APPROVED** [review]
I would squash the three commits into one and use the more general `refactor: make `addPromise()` generic` commit message.

Otherwise LGTM

**Review by mhevery — APPROVED** [review]

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:299433375:BASE:299433405:1583532599534:8d692e28)

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:299433375:BASE:299433840:1583532719480:8c63d75)

**Comment by matsko** [comment]
updated presubmit = https://test.corp.google.com/ui#id=OCL:299433375:BASE:299433840:1583770957489:2e3bf4af

**Comment by matsko** [comment]
@danranVm please rebase so we can see if this works with TypeScript 3.8.

**Comment by IgorMinar** [comment]
fyi: I just relabeled this PR as "master-only" since it's a "feature" that does not belong on the patch branch.

**Comment by danranVm** [comment]
> @danranVm please rebase so we can see if this works with TypeScript 3.8.

@matsko done

**Comment by AndrewKushnir** [comment]
Hi @danranVm, thanks for rebasing this PR. I've noticed that there is a new comment that was added as a result of rebase (`refactor: rebase code`). Could you please squash commits, so there is just the first one? Thank you.

**Comment by danranVm** [comment]
@AndrewKushnir ok,  it's correct now ?

**Comment by AndrewKushnir** [comment]
Hi @danranVm, thanks for updating commit messages, everything looks good now. I've started a [new g3 presubmit](http://test/OCL:301492772:BASE:301492785:1584493185980:a1ad15bb) and will keep you updated. Thank you.

**Comment by AndrewKushnir** [comment]
@danranVm, sorry, just noticed that the commit message refers to `addPromise` function, but the actual function being updated in this PR is `isPromise`. Also, the commit description uses `refactor:` namespace, when PR title says `feat(core)`. I'd propose to update commit message to something like this (also include *why* this change is made):

```
feat(core): add `isPromise` generic

This commit adds generic to `isPromise` function to help with type inference.
```

Could you please update commit message?

Thank you.

**Comment by danranVm** [comment]
@AndrewKushnir updated, thanks for your review.

**Comment by AndrewKushnir** [comment]
Hi @danranVm, thanks for the updates. The commit message looks good and g3 presubmit is successful, so I'm adding the "merge" label back. I also updated the "PR target" to be master and patch, since we cut 9.1.0 RC.0 this morning and created a new patch branch for 9.1.x versions. Thank you.

**Pull request closed without merging**